### DISCUSSION TITLE
Store for typeORM

### DISCUSSION THREAD

**Discussion by VictoriqueMoe** [nfriedly/express-rate-limit#442] [Show and tell]
Hi, i created a store for the TypeORM lib.

this should, in theroy allow anyone to use any DB supported by typeORM. https://www.npmjs.com/package/typeorm-rate-limit-store

**Comment by nfriedly** [comment]
Cool! I'll get it listed in the docs sometime soon.

**Reply by nfriedly** [reply]
Done, it's on the list at https://express-rate-limit.mintlify.app/reference/stores !

**Comment by nfriedly** [comment]
I took a quick look at the code and saw that the store doesn't have a `prefix` field, and then I realized that we didn't update the [Create a Store](https://express-rate-limit.mintlify.app/guides/creating-a-store) guide when we added that, so that's on me. But the basic idea is that you accept an optional string `prefix`, prefix it to the keys in the database, and also expose it at `store.prefix`. This allows a user to set up multiple limits (e.g. 10 hits per minute and 60 per hour, or `/api` gets 100/hour, but `/reset-password` only gets 3/hour.)

(This last part, putting the value at `store.prefix` is for the [singleCount validation check](https://github.com/express-rate-limit/express-rate-limit/blob/70dc0932d60b12cb5c3027a60585915b6a031b8f/source/validations.ts#L145-L159) so that it knows a request is not being double-counted if the user configured two rate limits with different prefixes, but it _is_ being double counted if they have the same prefix or no prefix.)

**Reply by nfriedly** [reply]
PR to update the docs at https://github.com/express-rate-limit/express-rate-limit/pull/443

**Comment by nfriedly** [comment]
One last thing - the same [offer I made to other folks who have create express-rate-limit stores](https://github.com/express-rate-limit/express-rate-limit/discussions/374): would you like to move the repo to the express-rate-limit org, and have @gamemaker1 and I help support and maintain it? 

No pressure, I just wanted to make the offer.

**Comment by VictoriqueMoe** [comment]
> I took a quick look at the code and saw that the store doesn't have a `prefix` field, and then I realized that we didn't update the [Create a Store](https://express-rate-limit.mintlify.app/guides/creating-a-store) guide when we added that, so that's on me. But the basic idea is that you accept an optional string `prefix`, prefix it to the keys in the database, and also expose it at `store.prefix`. This allows a user to set up multiple limits (e.g. 10 hits per minute and 60 per hour, or `/api` gets 100/hour, but `/reset-password` only gets 3/hour.)
> 
> (This last part, putting the value at `store.prefix` is for the [singleCount validation check](https://github.com/express-rate-limit/express-rate-limit/blob/70dc0932d60b12cb5c3027a60585915b6a031b8f/source/validations.ts#L145-L159) so that it knows a request is not being double-counted if the user configured two rate limits with different prefixes, but it _is_ being double counted if they have the same prefix or no prefix.)

Ah, ok. i will add the prefix,. i did see it, but i assumed it was something else. thanks 




> One last thing - the same [offer I made to other folks who have create express-rate-limit stores](https://github.com/express-rate-limit/express-rate-limit/discussions/374): would you like to move the repo to the express-rate-limit org, and have @gamemaker1 and I help support and maintain it?
> 
> No pressure, I just wanted to make the offer.

Is ok, i can maintain it,


but attest to that, i kinda find it odd how the scope of responsibility of clearing the entries from the store is passed down to the store itself. when i was wiriting the store. i expected the lib to call `deincrement` when the increment or `get` in the store returns the `resetTime` back to the caller and that timer has expired.

I was suprised to see that the responsiblity to actually clear out the entries in the store was down to the store itself. 

I expected `increment` to return an object, one key is `resetTime`, and the lib uses this `resetTime` to then call `deincrement` when the time is up. and when the hitcount is 0, it would call a clear method. I was wondering why my hit count was never going down, until i saw the default implementation has a `setInterval` on the init method. 

I thought to my self, why am i returning the `resetTime`, when the lib doesn't seem to do anything with it in terms of calling the store.

Maybe it would be better to make it clear that yes, you DO return `resetTime` but this isn't actually used as a timer internally. 

but as a other argument, it could be said that the responsiblity of keeping track of the timer should be the store due to clustered or federated node instances.

**Reply by nfriedly** [reply]
Thanks, I'll try and make some improvements to make it more clear.

**Comment by VictoriqueMoe** [comment]
I have created a PR for the store to include the prefix. if you can check and ensure that i have understood correctly, i would apppreicate it <3 

https://github.com/VictoriqueMoe/typeorm-rate-limit-store/pull/6/files

**Reply by nfriedly** [reply]
Cool, I'll take a look! It might take me a day or two to get to it, though.

**Reply by gamemaker1** [reply]
I reviewed the PR, you have used `prefix` correctly 🎉

**Reply by VictoriqueMoe** [reply]
thanks guys!
